### DISCUSSION TITLE
Particle grading with user-supplied function dependencies

### DISCUSSION THREAD

**Discussion by dion-w** [pybamm-team/PyBaMM#1729] [Q&A] [answered]
Hello everyone,
right now i am experimenting with a particle- (R(x)) and an active material volume fraction gradient (eps_am(x)) in the simulations. It's cool that there is a feature built in with the positional dependencies of both these parameters but the parameters naturally depending on it, like for example the electrical conductivity (sigma) are only depending on the temperature. One way of tackling this problem would be to alter the source code, and add the functional dependency of R and eps_am to the electrical conductivity. Since R and eps_am are depending on the position x in the electrode, it would also be possible to make the electrical conductivity depending on just x (sigma(x)). Unfortunately this would only be possible by changing the source code which i, for now, want to avoid as much as possible .

This was just an example. My question therefore is the following: is it possible to design a more general programmable interface with user-supplied dependencies of functions? PyBaMM could be even more flexible with this option. Is there anything considered so far?

Also, what confuses me: The active material volume fraction can vary in time but the porosity is not set up as a functional parameter depending on x. I saw that in the code you are calculating the inactive material by 1-eps_am-eps_pore. If you consider a changing eps_am wouldn't actually the porosity decrease first, before the inactive material gets compressed?

Best regards,
Dion Wilde

**Comment by brosaplanella** [comment, marked as answer]
About the first one, I don't think there is a workaround apart from modifying the source code. It should be made a function of `x`, and any other dependence (like R) can be introduced in the definition of the parameter.

I am not sure I understand the second question. The porosity is indeed a function of `x` and it can be defined as such. Even if the material gets compressed, the volume fractions of pore, active and inactive materials will always add up to 1.
